google/resource_storage_bucket.go

```diff
@@ -382,3 +453,130 @@ func flattenCors(corsRules []*storage.BucketCors) []map[string]interface{} {
 	}
 	return corsRulesSchema
 }
+
+func resourceGCSBucketLifecycleCreateOrUpdate(d *schema.ResourceData, sb *storage.Bucket) error {
+	if v, ok := d.GetOk("lifecycle_rule"); ok {
+		lifecycle_rules := v.([]interface{})
+
+		len_lifecycle_rules := len(lifecycle_rules)
+		if len_lifecycle_rules > 100 {
+			return fmt.Errorf("At most 100 lifecycle_rule blocks are allowed, %d given", len_lifecycle_rules)
+		}
+
+		sb.Lifecycle = &storage.BucketLifecycle{}
+		sb.Lifecycle.Rule = make([]*storage.BucketLifecycleRule, 0, len_lifecycle_rules)
+
+		for _, lifecycle_rule := range lifecycle_rules {
+			lifecycle_rule := lifecycle_rule.(map[string]interface{})
+
+			target_lifecycle_rule := &storage.BucketLifecycleRule{}
+
+			if v, ok := lifecycle_rule["action"]; ok {
+				action := v.(*schema.Set).List()[0].(map[string]interface{})
+
+				target_lifecycle_rule.Action = &storage.BucketLifecycleRuleAction{}
+
+				if v, ok := action["type"]; ok {
+					target_lifecycle_rule.Action.Type = v.(string)
+				}
+
+				if v, ok := action["storage_class"]; ok {
+					target_lifecycle_rule.Action.StorageClass = v.(string)
+				}
+			}
+
+			if v, ok := lifecycle_rule["condition"]; ok {
+				condition := v.(*schema.Set).List()[0].(map[string]interface{})
+
+				if len(condition) < 1 {
+					return fmt.Errorf("At least one condition element is required")
```

Comment:
I think I remember why I implemented this check differently at first by incrementing a counter. From the XML API doc:

> Condition	Defines the condition under which the action will be taken, which must contain *at least one specific condition element*. Required.

cf. https://cloud.google.com/storage/docs/xml-api/put-bucket-lifecycle#request_body_elements